doc/adr/0005-jwt-authentication.md

```diff
@@ -0,0 +1,171 @@
+# 5. JWT Authentication Extensibility
+
+Date: 2020-07-01
+
+## Status
+
+Proposed.
+
+
+## Context
+
+We wish to have the option to, in future, extend or change the authentication
+flow in a way which clients can safely handle.  This architecture design note
+provides for a check which clients can safely add now, to be compatible with
+such future changes and to be proof against certain attacks.
+
+Upon connection, the NATS server sends an `INFO` line with various fields; the
+NATS client processes that, performs a TLS upgrade if appropriate, and then
+sends a `CONNECT` line with various fields.
+
+At present, if a server supports JWT authentication then it generates a strong
+random nonce, presented as the `nonce` field of the `INFO` line.  If a client
+wishes to authenticate with a JWT, then it takes the bytes of that field, and
+without any decoding signs them with the private key associated with the jwt;
+the resulting signature is base64-encoded; then in the `CONNECT` line the
+client includes two fields: `jwt` containing their JWT, and `sig` which
+contains the base64-encoded result of signing the nonce.
+
+This protects the server against replay attacks, since a strongly random nonce
+is chosen each time.  It provides no inherent protection against
+man-in-the-middle attacks, relying instead upon TLS checks.  The server can
+enforce mutual-TLS, but not all deployment scenarios encoforce this.
+
+A well-behaved client with a competent TLS stack, configured to only proceed
+with TLS, will verify the NATS server's certificate before proceeding, so will
+only send the signature to a server which can present a certificate that
+client expects.
```

Comment:
If I understand the problem correctly, we could envision forcing the server to limit JWT nonce signing to clients that have TLS and verify client certs, but that does not protect clients that are configured to NOT verify server certificate. That being said, if server requires client cert and verify it, would not the agent in the middle have to already provide valid certs?

Reply:
If the server requires a client cert and verifies it, then that is complete authentication and no in-protocol authentication is needed, no JWTs.  There is effectively going to be structured nonce signing by the TLS handshake.  So _if_ mutual TLS is an available solution then it works.  Unless we're planning to move to a world of "only mTLS" and abandoning the JWT user model, that doesn't help here, where the core problem is that today's unstructured signers will be easy to trick into signing any future structured data.